internal/cms/signed.go

```diff
@@ -0,0 +1,313 @@
+	}
+	opts.Intermediates = intermediates
+	verifiedSignerMap := map[string]*x509.Certificate{}
+	var lastErr error
+	for _, signerInfo := range d.SignerInfos {
+		signingCertificate := d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			lastErr = ErrCertificateNotFound
+			continue
+		}
+
+		cert, err := d.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, cert.Raw)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = cert
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertificates := make([]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, cert := range verifiedSignerMap {
+		verifiedSigningCertificates = append(verifiedSigningCertificates, cert)
+	}
+	return verifiedSigningCertificates, nil
+}
+
+// VerifySigner verifies the signerInfo against the user specified signingCertificate.
+//
+// This function should be used when:
+//
+// 1. The certificates field of d is missing. This function allows the caller to provide
+// a signing certificate to verify the signerInfo.
+//
+// 2. The caller doesn't trust the signer identifier (unsigned field) of signerInfo
+// to identify signing certificate. This function allows such caller to use their trusted
+// signing certificate.
+//
+// Note: the intermediate certificates (if any) and root certificates in the verify
+// options MUST be set by the caller. The certificates field of d is not used in this function.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) VerifySigner(ctx context.Context, signerInfo *SignerInfo, signingCertificate *x509.Certificate, opts x509.VerifyOptions) (*x509.Certificate, error) {
+	if signerInfo == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signer info is required"}
+	}
+
+	if signingCertificate == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signing certificate is required"}
+	}
+
+	if signerInfo.Version != 1 {
+		// Only IssuerAndSerialNumber is supported currently
+		return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+	}
+
+	return d.verify(signerInfo, signingCertificate, &opts)
+}
+
+// verify verifies the trust in a top-down manner.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verify(signerInfo *SignerInfo, cert *x509.Certificate, opts *x509.VerifyOptions) (*x509.Certificate, error) {
+	// verify signer certificate
+	certChains, err := cert.Verify(*opts)
```

Comment:
@priteshbandi This is done in notation-core-go as the requirements of certificate chain is Notary Project specific. cms/timestamp would form a certificate chain on verification and return it to its caller.

Reply:
I am worried about the use outside notation. User will assume that verify operation is validating that the cert chain is correct.

Reply:
> I am worried about the use outside notation. User will assume that verify operation is validating that the cert chain is correct.

Makes sense. `cert.Verify` would only return a valid certificate chain on success. Otherwise, an error will be returned. @JeyJeyGao to confirm this.

Reply:
https://github.com/JeyJeyGao/tspclient-go/blob/a872da4201c3ff61b09005aff4e72b3951525e44/internal/cms/signed.go#L193-L198

Yes, by calling cert.Verify(opts), we can ensure that the returned certificate chain is valid.

Reply:
Updated my comment to include time-stamping requirement

Reply:
> I agree that the validation of notation-specific requirements should be conducted within 'notation-core-go'. However, it's we should ensure that the timestamping certificate(s) adheres to the CA/Browser Forum Baseline Requirements (BR).
> 
> https://github.com/cabforum/code-signing/blob/v3.7/docs/CSBR.md
> 
> It should be integrated into the verify operation bcoz callers of the function will expect library to perform fundamental checks.

@priteshbandi RFC 3161 related will be in PR #18 including certificate validation.